### ISSUE TITLE
Last-written-LSN calculation problem

### ISSUE THREAD

**Issue by knizhnik** [neondatabase/neon#7518]
## Steps to reproduce

Working on move of two-phase index build to neon extension https://github.com/neondatabase/neon/pull/7440
I was faced with some strange errors, which can be reproduced with small modification on main branch.
What this PR originally tries to do - is to read page from PS and if it is zero page, then read it from local storage.

I have added the similar code to main branch:
```
diff --git a/pgxn/neon/pagestore_smgr.c b/pgxn/neon/pagestore_smgr.c
index 57a16e00c..efcc9153d 100644
--- a/pgxn/neon/pagestore_smgr.c
+++ b/pgxn/neon/pagestore_smgr.c
@@ -2257,8 +2257,15 @@ neon_read(SMgrRelation reln, ForkNumber forkNum, BlockNumber blkno, void *buffer
                case RELPERSISTENCE_PERMANENT:
                        break;
 
-               case RELPERSISTENCE_TEMP:
                case RELPERSISTENCE_UNLOGGED:
+                       if (forkNum == MAIN_FORKNUM &&
+                               unlogged_build_rel == reln &&
+                               (unlogged_build_phase == UNLOGGED_BUILD_PHASE_1 || unlogged_build_phase == UNLOGGED_BUILD_PHASE_2))
+                       {
+                               request_lsn = neon_get_request_lsn(&latest, InfoFromSMgrRel(reln), forkNum, blkno);
+                               neon_read_at_lsn(InfoFromSMgrRel(reln), forkNum, blkno, request_lsn, latest, buffer);
+                       }
+               case RELPERSISTENCE_TEMP:
                        mdread(reln, forkNum, blkno, buffer);
                        return;
```

and got similar errors, for example:
```
2024-04-25T18:39:04.223443Z ERROR page_service_conn_main{peer_addr=[::1]:58738}:process_query{tenant_id=ae10947c5361833e9fc8f7ae373f310e timeline_id=4ea450df431da0ff14f6690535e36ed3}:handle_pagerequests:handle_get_page_at_lsn_request{rel=1663/16384/33234 blkno=0 req_lsn=0/204DA58}: error reading relation or page version: Read error: could not find data for key 000000067F00004000000081D200FFFFFFFF (shard ShardNumber(0)) at LSN 0/1514AD8, for request at LSN 0/11A7DC50
```

I noticed that creation of relation. was successfully wal-logged:
```
PG:2024-04-25 18:39:04.216 GMT [49324] LOG:  XLOG_SMGR_CREATE at 0/11A7DC80
PG:2024-04-25 18:39:04.216 GMT [49324] STATEMENT:  CREATE INDEX jidx ON testjsonb USING gin (j jsonb_path_ops);
```
... but at more recent LSN than requested.

Looking at code of `log_smgrcreate`:
```
void
log_smgrcreate(const RelFileLocator *rlocator, ForkNumber forkNum)
{
	xl_smgr_create xlrec;
	XLogRecPtr lsn;

	/*
	 * Make an XLOG entry reporting the file creation.
	 */
	xlrec.rlocator = *rlocator;
	xlrec.forkNum = forkNum;

	XLogBeginInsert();
	XLogRegisterData((char *) &xlrec, sizeof(xlrec));
	lsn = XLogInsert(RM_SMGR_ID, XLOG_SMGR_CREATE | XLR_SPECIAL_REL_UPDATE);
	elog(LOG, "XLOG_SMGR_CREATE %d at %X/%X", rlocator->relNumber, LSN_FORMAT_ARGS(lsn));
	SetLastWrittenLSNForRelation(lsn, *rlocator, forkNum);
}
```
... I noticed that it is updating last written LSN for relation size.

But when we request page of the relation we perform lookup in LW_LSN hash using different key (including block number) and if it is not found then use global `maxLastWrittenLsn`. But it can be smaller than LSN of XLOG_SMGR_CREATE!

Why we didn't faced with this problem before?
I think it is because Postgres is always first check relation size (`smgrblocks`) before actually accessing its pages.

But in this case it doesn't work because relation size is cached locally and so no get_rel_size request was sent to PS.

Can we rely on this two assumptions:
- smgrnblocks is always called before getting relation page
- relation size was not cached

Or should we slightly change the way of how LW-LSB is updated.

For example the following small patch fixes the problem (I was not able to reproduce it any more):
```
diff --git a/src/backend/access/transam/xlog.c b/src/backend/access/transam/xlog.c
index 9f343cb412..10dcc33edb 100644
--- a/src/backend/access/transam/xlog.c
+++ b/src/backend/access/transam/xlog.c
@@ -6354,6 +6354,10 @@ SetLastWrittenLSNForBlockRange(XLogRecPtr lsn, RelFileLocator rlocator, ForkNumb
                        else
                        {
                                entry->lsn = lsn;
+
+                               if (key.blockNum == REL_METADATA_PSEUDO_BLOCKNO && lsn > XLogCtl->maxLastWrittenLsn)
+                                       XLogCtl->maxLastWrittenLsn = lsn;
+
                                if (hash_get_num_entries(lastWrittenLsnCache) > lastWrittenLsnCacheSize)
                                {
```
So of relation metadata is updated, then make sure that global `lastWrittenLsnCache` is not smaller than this LSN.
In other words: LSN of requesting relation page can not be smaller than LSN of relation metadata.

Should I create PR for it?

## Logs, links
-

**Comment by hlinnaka** [comment]
To summarize, the sequence of smgr calls is:

1. smgrcreate()
2. smgrwrite(blk 0)
3. smgrread(blk 0)

With normal logged relations, the smgrwrite() function updates the last-written LSN for block 0, but with your patch, that's not happening.

**Comment by knizhnik** [comment]
Yes, you are right.
But I wonder if the following scenario can happen even without my PR:
1. smgrcreate
2. smgrextend (doesn't update last written LSN)
3. smgrread(0)

**Comment by hlinnaka** [comment]
smgrextend() does update the last-written LSN.

I don't see a bug here on 'main', closing.
